### PR TITLE
Go modules migration

### PR DESCRIPTION
**Pull request by eikenb**
Change over to go modules (without vendor) for dependency management.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -6,4 +6,13 @@ jobs:
     working_directory: /go/src/github.com/hashicorp/envconsul
     steps:
       - checkout
-      - run: go test -v
+      - restore_cache:
+          keys:
+            - ct-modcache-v1-{{ checksum "go.mod" }}
+      - run: |
+          export GO111MODULE=on
```
Comment (mikemorris) [inline .circleci/config.yml]: This could be pulled up into an [`environment`](https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-job) block in CircleCI config to be a little cleaner.

**Comment by eikenb** [comment]
@mkeeler Yes, I used vendor to speed up the docker based builds. Everything built fine without it, just annoyingly slow. I didn't think of just adding a bind mount to my mod-cache and that would accomplish the same thing. Though I can think of pros/cons each way and am not sure which is best. The pros for vendor are not requiring the (now optional) GOPATH and being more explicit. The pros for the bind mount are that it won't conflict with a downstream or other use of vendor and it keeps the Makefile a bit cleaner. Hmm.

I decided against keeping the vendor directory as the hashicorp go-modules RFC specifies no vendor directory and most people seem to just grab the pre-built binaries anyways. If it is enough of an issue for someone they will file a bug and I'd reconsider it at that point.

File: .circleci/config.yml (hunk 1 above)
Comment (eikenb) [reply .circleci/config.yml]: That is cleaner. I wasn't thinking I'd need it long due to it not being required with go1.13, so I didn't give it much thought. I'll move it.

**Comment by mkeeler** [comment]
@eikenb I think both the build time vendoring or modcache mapping should work about the same. The big difference is that with mapping the modcache it wont touch the local source so there is nothing to clean up (or accidentally commit back to the repo). 

The vendor dir being gone seems fine to me.

File: .circleci/config.yml (hunk 1 above)
Comment (eikenb) [reply .circleci/config.yml]: One note about this. You have to quote "on" in the environment setting section, otherwise circleci converts it to "true".

**Comment by eikenb** [comment]
@mkeeler .. I tried switching to use volume/bind mounts to see how it'd look and realized a problem. I use the build-time vendor directory not only for the binaries, but to build the docker images as well. In the case of the latter it uses `go build` which doesn't take a volume/mount argument and so would still either download things or need a copy (ie. vendor). 

TLDR; I think I'll just stick with the build-time vendor setup I have now.

**Comment by eikenb** [comment]
Scratch that. Going with a hybrid. Realized I can't just run `make build` right now as it requires the vendor directory. So I'm going to keep the mod vendor setup for building docker images but mount the local mod-cache for building binaries. 

I'll restrict the `go mod vendor` and `rm -rf vendor` calls to the build-docker target and add vendor/ to .gitignore to help lessen the downsides of having both.
